File: Python/programmers/etc/level2/숫자_변환하기.py
from collections import deque

def solution(x, y, n):
    q = deque([x])
    visited = [-1]*(y+1)
    visited[x] = 0
    while q:
        cur = q.popleft()
        for nxt in [cur+n, cur*2, cur*3]:
            if nxt <= y and visited[nxt] == -1:
                visited[nxt] = visited[cur]+1
                q.append(nxt)
    return visited[y]


x, y, n = 10, 40, 5
print(solution(x, y, n)) # 2

x, y, n = 10, 40, 30
print(solution(x, y, n)) # 1

x, y, n = 2, 5, 4
print(solution(x, y, n)) # -1

File: Java/programmers/etc/level2/숫자변환하기.java
package CodeTest.Java.programmers.etc.level2;

import java.util.*;

public class 숫자변환하기 {
    public int solution(int x, int y, int n) {
        Queue<Integer> q = new LinkedList<>();
        q.add(x);
        int[] visited = new int[y+1];
        Arrays.fill(visited, -1);
        visited[x] = 0;
        
        while (!q.isEmpty()) {
            int cur = q.poll();
            int[] tmp = new int[]{cur+n, cur*2, cur*3};
            for (int nxt : tmp) {
                if (nxt <= y && visited[nxt] == -1) {
                    visited[nxt] = visited[cur]+1;
                    q.add(nxt);
                }
            }
        }
        
        return visited[y];
    }
}
